Translate 1 file to ko - Understanding Errors (#129)
microsoft/TypeScript-Website-Localizations: docs/documentation/ko/handbook-v2/Understanding Errors.md

---
title: Understanding Errors
layout: docs
permalink: /ko/docs/handbook/2/understanding-errors.html

> oneline: "타입스크립트에서 오류들을 읽는법."

Reviewer: ```suggestion
oneline: "TypeScript에서 오류를 보는 방법."
```
굳이 복수로 안써도 될 것 같아요!

근데 "읽는다" 보다 "보는 방법"은  어떤가요?

타입스크립트는 TypeScript로 변경해주세요!

Author: @bumkeyy 

수정 완료했습니다~!!

---

> # 오류를 이해하기

Reviewer: ```suggestion
# 오류를 이해하기 (Understanding Errors)
```

TypeScript는 오류를 찾으면, 무엇이 잘못됐는지 가능한 자세하게 설명하려고 합니다.

> TypeScript의 타입 시스템은 구조적이기 때문에, 이것은 종종 문제를 발견한 장소에 대한 다소 긴 설명을 제공한다는 것을 의미합니다.

Reviewer: ```suggestion
TypeScript의 타입 시스템은 구조적이기 때문에, 문제를 발견한 위치에 대해 자세한 설명을 제공할 수 있습니다.
```
[제안]

> ## 용어

Reviewer: ```suggestion
## 용어 (Terminology)
```

> 오류 메시지에서 자주 볼 수 있는 몇 가지 용어가 있어 이해하는 데 도움이 될 것입니다.

Reviewer: ```suggestion
오류 메시지에서 자주 등장하는 용어를 알면 이해하는 데 도움이 됩니다.
```
[제안]

> #### _할당할 수 있는_ (_assignable to_)

Reviewer: ```suggestion
#### _assignable to_
```
이 부분은 실제 오류메시지의 예시라서
원문 그대로 사용해야 할 것 같아요! 😄

> TypeScript는 _할당할 수 있는_ 다른 형식(하나의 형식을 대체할 수 있는 경우)을 고려합니다.

Reviewer: ```suggestion
TypeScript는 타입이 다른 타입으로 대체할 수 있을 때 타입을 다른 타입에 _할당할 수 있다_ 라고 표현합니다.
```
흠 실제 메시지를 번역하려니 조금 어색하네요 😓

> 다시 말해 `고양이`는 `동물`을 대신할 수 있는 존재이기 때문에 `동물`에게 _할당할 수 있는_ 존재입니다.

Reviewer: ```suggestion
다시 말해 `고양이`는 `동물`을 대체할 수 있기 때문에 `동물`에게 _할당할 수_ 있습니다.
```

> 이 관계는 이름에서 알 수 있듯이, `t`와 `s`의 유형을 조사하여 할당 `t = s;`의 타당성을 확인하는 데 사용됩니다.

Reviewer: ```suggestion
이름에서 보이듯, 이런 관계는 `t`와 `s`의 타입을 검사하여 `t = s;`의 할당 타당성을 확인하는 데 사용됩니다.
```
[제안]

> 또한 두 가지 유형이 상호 작용하는 대부분의 다른 위치를 확인하는 데도 사용됩니다.

Reviewer: ```suggestion
또한 두 가지 타입이 상호 작용하는 대부분의 위치에서 확인할 때에도 사용됩니다.
```

Reviewer: [제안]

> 예를 들어, 함수를 호출할 때 각 인수의 유형은 매개 변수의 선언된 유형에 _할당할 수_ 있어야합니다.

Reviewer: ```suggestion
예를 들어, 함수를 호출할 때 각 인수의 타입은 매개 변수의 선언된 유형에 _할당할 수_ 있어야합니다.
```

> 비공식적으로 `T는 S에 할당할 수 없습니다`라고 하면 TypeScript는 "_`T`와 `S`는 호환되지 않는다"_.고 말한다고 생각하면됩니다.

Reviewer: ```suggestion
비공식적으로 `T is not assignable to S`라고 하면 TypeScript는 "_`T`와 `S`는 호환되지 않는다"_.고 말한다고 생각하면됩니다.
```
실제 메시지는 원문으로 쓰는게 좋을 것 같아요 😄

> 그러나, 이것은 _방향성이 있는_ 관계라는 점에 유의하십시오: `S`가 `T`에 할당될 수 있다고 해서 `T`가 `S`에 할당될 수 있는 것은 아닙니다.

Reviewer: ```suggestion
그러나, 이것은 _방향성이 있는_ 관계라는 점에 유의하세요: `S`가 `T`에 할당될 수 있다고 해서 `T`가 `S`에 할당될 수 있는 것은 아닙니다.
```

Reviewer: [제안]

## 예시들 (Examples)

오류 메시지의 몇 가지 예제를 살펴보고 어떤 일이 일어나고 있는지 알아보겠습니다.

> ### 오류 정교화

Reviewer: ```suggestion
### 오류 정교화 (Error Elaborations)
```

각 오류는 선행 메시지로 시작하고 때로는 더 많은 하위 메시지로 이어집니다.
각 하위 메시지는 위의 메시지에 대한 "왜?" 질문에 대답하는 것으로 생각할 수 있습니다.
몇 가지 예를 통해 실제로 어떻게 작동하는지 살펴보겠습니다.

다음은 예제 자체보다 긴 오류 메시지를 생성하는 예입니다:

```ts twoslash
// @errors: 2322
let a: { m: number[] }
let b = { m: [""] }
a = b
```

> 마지막 줄을 확인하는 동안 TypeScript에서 오류가 발생했습니다.

Reviewer: ```suggestion
마지막 줄에서 TypeScript는 오류를 발견했습니다.
```
[제안]

오류 발생에 대한 논리는 할당이 정상인지 확인하는 논리에서 비롯됩니다:

1. `b` 타입은 `a` 타입에 할당 가능한가요? 아뇨. 왜요?

> 2. 왜냐하면 `m` 속성의 유형이 호환되지 않기 때문입니다. 왜죠?

Reviewer: ```suggestion
2. 왜냐하면 `m` 속성의 타입이 호환되지 않기 때문입니다. 왜죠?
```

3. 왜냐하면 `b`의 `m` 속성(`string[]`)은 `a`의 `m` 속성(`number[]`)에 할당할 수 없기 때문입니다. 왜죠?

> 4. 한 배열의 요소 유형(`string`)을 다른 배열(`number`)에 할당할 수 없기 때문입니다.

Reviewer: ```suggestion
4. 한 배열의 요소 타입(`string`)을 다른 타입(`number`)에 할당할 수 없기 때문입니다.
```
